ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeCallback.java

```diff
@@ -17,34 +17,80 @@
+   * exceed usage. For examples about when usage could be allowed to exceed quota, check {@link QuotaEnforcer}.
+   *
+   * If forceCharge is specified as {@code true}, then the charge must happen. In that case the implementations can skip
+   * any checks for allowing the request, as those checks are inconsequential.
+   *
+   * @param shouldCheckExceedAllowed if set to {@code true}, check if usage is allowed to exceed quota.
+   * @param forceCharge if set to {@code true}, then charge must happen. Implementation may decide to skip any checks as
+   *                    they are inconsequential.
+   * @param chunkSize of the chunk.
+   * @throws QuotaException In case of any exception.
+   * @return QuotaAction representing the recommended action to take for quota compliance.
+   */
+  QuotaAction checkAndCharge(boolean shouldCheckExceedAllowed, boolean forceCharge, long chunkSize) throws QuotaException;
+
+  /**
+   * Charges the cost of chunk request against quota, if checks to allow a chunk request are successful. The cost to
+   * charge is one unit of quota cost, as determined by {@link QuotaConfig#quotaAccountingUnit}.
+   * For more details check {@link QuotaChargeCallback#checkAndCharge(boolean, boolean, long)}.
+   *
+   * @param shouldCheckExceedAllowed if set to {@code true}, check if usage is allowed to exceed quota.
+   * @param forceCharge if set to {@code true}, then charge must happen. Implementation may decide to skip any checks as
+   *                    they are inconsequential.
+   * @throws QuotaException In case of any exception.
+   * @return QuotaAction representing the recommended action to take for quota compliance.
+   */
+  default QuotaAction checkAndCharge(boolean shouldCheckExceedAllowed, boolean forceCharge) throws QuotaException {
+    return checkAndCharge(shouldCheckExceedAllowed, forceCharge, getQuotaConfig().quotaAccountingUnit);
+  }
+
   /**
    * Callback method that can be used to charge quota usage for a request or part of a request.
    * @param chunkSize of the chunk.
    * @throws QuotaException In case request needs to be throttled.
    */
+  @Deprecated
   void charge(long chunkSize) throws QuotaException;
 
   /**
    * Callback method that can be used to charge quota usage for a request or part of a request. Call this method
    * when the quota charge doesn't depend on the chunk size.
    * @throws QuotaException In case request needs to be throttled.
    */
+  @Deprecated
   void charge() throws QuotaException;
 
   /**
    * Check if request should be throttled based on quota usage.
    * @return {@code true} if request usage exceeds limit and request should be throttled. {@code false} otherwise.
    */
+  @Deprecated
   boolean check();
 
   /**
    * Check if usage is allowed to exceed the quota limit.
    * @return {@code true} if usage is allowed to exceed the quota limit. {@code false} otherwise.
    */
+  @Deprecated
   boolean quotaExceedAllowed();
```

Comment:
we can just remove these methods. These methods are only used by ambry.

Reply:
To remove these methods, we will also have to change the `Chargeable` interface and its implementations. That would make this PR very large. As mentioned in the PR description, once all the changes are done, I will remove these methods.